core/src/main/scala/org/apache/spark/memory/MemoryManager.scala

```diff
@@ -255,7 +256,17 @@ private[spark] abstract class MemoryManager(
     }
     val size = ByteArrayMethods.nextPowerOf2(maxTungstenMemory / cores / safetyFactor)
     val default = math.min(maxPageSize, math.max(minPageSize, size))
-    conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    val sizeAsBytes = conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    // If we are using G1 GC, it's better to take the LONG_ARRAY_OFFSET into consideration
+    // so that the requested memory size is power of 2 and can be divided by G1 region size
+    // to reduce memory waste within one G1 region
+    if (Utils.isG1GarbageCollector &&
```

Comment:
I'm just curious whether this adjustment can be done in `HeapMemoryAllocator.allocate` instead? since it's closer to the actual logic.

Reply:
I prefer not to put it in `HeapMemoryAllocator.allocate` because it may break the semantics. When we do `HeapMemoryAllocator.allocate(size)` we expected to get memory with specified size or throw oom, but we internally change it to another value `(size-Platform.LONG_ARRAY_OFFSET)`, which may crash the caller's code or brings confusion